### PR TITLE
perf: Improve dashboard performance by decreasing rerenders

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
SUPERSEDED BY PRs: https://github.com/apache/superset/pull/31240, https://github.com/apache/superset/pull/31241, https://github.com/apache/superset/pull/31242, https://github.com/apache/superset/pull/31243, https://github.com/apache/superset/pull/31244.
This PR is not to be merged - aggregates 5 smaller PRs to be referenced as total performance improvement.

Due to many inefficient redux selectors and other factors causing numerous rerenders, large dashboards were running very slow.

This PR refactors many components related to Dashboard and native filters - some components are rewritten to functional to utilize React's memoization, Redux selectors optimized to select smaller objects that change less frequently, optimized frequently called functions.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

The tests were conducted on a large dashboard - 300+ charts, 40+ filters. Test steps: reload the page, let the 1st tab load, switch to a different tab, let everything load.
Tested on my M2 Pro Macbook - since it's likely that an average Superset user works on a slower computer, I also run the tests with 4x and 20x CPU throttling.
Testing tool - chrome dev tools, performance tab

1. No throttling
Before - total scripting time 2.73s (screenshot 1), each chart holder blocks the main thread for ~70ms
After - total scripting time 1.48s (screenshot 2), each chart holder blocks the main thread for ~5ms

2. 4x throttling
Before - total scripting time 11.93s (screenshot 3), each chart holder blocks the main thread for ~300ms
After - total scripting time 6.23s (screenshot 4), each chart holder blocks the main thread for ~20ms

3. 20x throttling
Before - total scripting time 82s (screenshot 5), each chart holder blocks the main thread for ~2s
After - total scripting time 32s (screenshot 6), each chart holder blocks the main thread for ~100ms

**SCREENSHOTS BEFORE:**

1. No throttling

image

2. 4x throttling

image

3. 20x throttling

image

**SCREENSHOTS AFTER:**

1. No throttling

image

2. 4x throttling

image

3. 20x throttling

image

### TESTING INSTRUCTIONS

No functional changes - all dashboard functionalities should work like before.

Recommend thorough testing of:
- Filter interactions
- Chart loading/rendering
- Dashboard navigation
- Cross-filter functionality
- Dashboard state persistence

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by villebro** [comment]
These PRs from Kamil always make me soo happy! ❤️

**Review by villebro — COMMENTED** [review]
I did my best to find anything to comment or even nit about in this PR, but this looks really solid. I will have to resort to testing to see there's anything that looks out of the ordinary. Will approve once I'm done with the testing rounds..

**Comment by kgabryje** [comment]
@villebro  unfortunately the ephemeral environments have been broken for some time now 🙁

File: superset-frontend/src/dashboard/components/gridComponents/Chart.jsx [hunk 1]
```diff
@@ -32,13 +34,30 @@ import {
...
+  addDangerToast,
+  addSuccessToast,
+} from '../../../components/MessageToasts/actions';
+import {
+  setFocusedFilterField,
+  toggleExpandSlice,
+  unsetFocusedFilterField,
+} from '../../actions/dashboardState';
+import { changeFilter } from '../../actions/dashboardFilters';
+import { refreshChart } from '../../../components/Chart/chartAction';
```
Comment (geido) [inline superset-frontend/src/dashboard/components/gridComponents/Chart.jsx:53]: Can we use absolute paths?

**Review by geido — APPROVED** [review]
Thank you so much for this! Code looks good but it's a gigantic PR that needs more than a first-pass code review. I'd suggest splitting this into smaller pieces if possible to help with functional tests.

**Pull request closed without merging**